File: android/src/main/java/com/assaabloy/plugins/permissions/PermissionsPlugin.java
package com.assaabloy.plugins.permissions;

import android.Manifest;

import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(
    name = "Permissions",
    permissions = {
        @Permission(alias = "microphone", strings = { Manifest.permission.RECORD_AUDIO })
    }
)
public class PermissionsPlugin extends Plugin {

    @PluginMethod
    public void query(PluginCall call) {
        String permission = call.getString("name");

        if (permission == null) {
            call.reject("Permission name is required");
            return;
        }

        switch (permission) {
            case "microphone":
                checkMicrophone(call);
                break;
            default:
                call.reject("Unknown permission");
        }
    }

    @PluginMethod
    public void request(PluginCall call) {
        String permission = call.getString("name");

        if (permission == null) {
            call.reject("Permission name is required");
            return;
        }

        switch (permission) {
            case "microphone":
                requestMicrophone(call);
                break;
            default:
                call.reject("Unknown permission");
        }
    }

    private void checkMicrophone(PluginCall call) {
        PermissionState state = getPermissionState("microphone");

        if (state == PermissionState.GRANTED) {
            call.resolve(createResult(PermissionState.GRANTED));
        } else if (state == PermissionState.PROMPT) {
            call.resolve(createResult(PermissionState.PROMPT));
        } else {
            call.resolve(createResult(PermissionState.DENIED));
        }
    }

    private void requestMicrophone(PluginCall call) {
        if (getPermissionState("microphone") == PermissionState.GRANTED) {
            call.resolve(createResult(PermissionState.GRANTED));
        } else {
            requestPermissionForAlias("microphone", call, "microphonePermissionCallback");
        }
    }

    @PermissionCallback
    private void microphonePermissionCallback(PluginCall call) {
        PermissionState state = getPermissionState("microphone");

        if (state == PermissionState.GRANTED) {
            call.resolve(createResult(PermissionState.GRANTED));
        } else {
            call.resolve(createResult(PermissionState.DENIED));
        }
    }

    private JSObject createResult(PermissionState state) {
        JSObject result = new JSObject();
        result.put("status", state.toString().toLowerCase());
        return result;
    }
}

File: ios/Sources/PermissionsPlugin/PermissionsPlugin.swift
import Foundation
import Capacitor
import OSLog
import Network
import CoreLocation
import AVFoundation


@objc(PermissionsPlugin)
public class PermissionsPlugin: CAPPlugin, CAPBridgedPlugin, NetServiceDelegate {
    public var identifier = "PermissionsPlugin"
    public var jsName = "Permissions"
    public var pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "query", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "request", returnType: CAPPluginReturnPromise)
    ]

    @objc func query(_ call: CAPPluginCall) {
        let permission = call.getString("name")

        if permission == nil {
            call.reject("Permission name is required")
            return
        }
        
        switch permission {
        case "microphone":
            checkMicrophone(call)
        default:
            call.reject("Unknown permission")
        }
    }

    @objc func request(_ call: CAPPluginCall) {
        let permission = call.getString("name")

        if permission == nil {
            call.reject("Permission name is required")
            return
        }
        
        switch permission {
        case "microphone":
            requestMicrophone(call)
        default:
            call.reject("Unknown permission")
        }
    }
  
    @objc func checkMicrophone(_ call: CAPPluginCall) {
        let status = AVCaptureDevice.authorizationStatus(for: .audio)
        
        switch status {
        case .authorized:
            call.resolve(["status": "granted"])
        case .denied:
            call.resolve(["status": "denied"])
        case .notDetermined:
            call.resolve(["status": "notDetermined"])
        case .restricted:
            call.resolve(["status": "restricted"])
        @unknown default:
            call.resolve(["status": "unknown"])
        }
    }

    @objc func requestMicrophone(_ call: CAPPluginCall) {
        AVCaptureDevice.requestAccess(for: .audio) { granted in
            if granted {
                call.resolve(["status": "granted"])
            } else {
                call.resolve(["status": "denied"])
            }
        }
    }
}
